### ISSUE TITLE
no-extra-semi warns about unnecessary semicolon in export statement

### ISSUE THREAD

**Issue by lo1tuma** [eslint/eslint#3184]
The `no-extra-semi` rule warns about the following code:

```js
export default function() {};
```

Testcase on current master:
```js
eslintTester.addRuleTest("lib/rules/no-extra-semi", {
    valid: [
        // ...
        {
            code: "export default function () {};",
            ecmaFeatures: { modules: true }
        }
        // ...
    ]
});
```

throws the following assertion error:

```
1) no-extra-semi export default function () {};:
     AssertionError: Should have no errors but had 1: [ { ruleId: 'no-extra-semi',
    severity: 1,
    message: 'Unnecessary semicolon.',
    line: 1,
    column: 30,
    nodeType: 'EmptyStatement',
    source: 'export default function () {};' } ]: expected 1 to equal 0
```

If I export a variable (e.g. `export default foo;`) instead of a function expression I got no error.

-----

Is this a bug in the parser because the semicolon after the export statement is represented as an `EmptySatement`?

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by btmills** [comment]
I did an analysis of this part of the spec back in [#2194 (comment)](https://github.com/eslint/eslint/issues/2194#issuecomment-88575851). Based on that, an `ExportDefaultDeclaration` whose `declaration` is a `FunctionDeclaration` (among other alternatives) does not have a semicolon. It looks like this is being parsed correctly as an export followed by an empty statement, but feel free to recheck my work - there's a source link in the comment.

**Comment by lo1tuma** [comment]
In my example I have a `FunctionExpression` instead of a `FunctionDeclaration`.

**Comment by lo1tuma** [comment]
Interesting, [14.1.3](http://www.ecma-international.org/ecma-262/6.0/#sec-function-definitions-static-semantics-boundnames) in the spec says an anonymous functions can be `HoistableDeclaration` if it’s used in a export declaration.
So you are probably right @btmills, that the rule is behaving correctly.

**Comment by btmills** [comment]
Parsing

```js
export default function () { };
```

Yields

```js
{ type: 'Program',
  body:
   [ { type: 'ExportDefaultDeclaration',
       declaration:
        { type: 'FunctionDeclaration',
          id: null,
          params: [],
          body: { type: 'BlockStatement', body: [] },
          generator: false,
          expression: false } },
     { type: 'EmptyStatement' } ],
  sourceType: 'module' }
```

It's weird, but it's an unnamed function declaration. For more discussion on it, see [estree/estree#36 (comment)](https://github.com/estree/estree/pull/38#issuecomment-76960163).

**Comment by lo1tuma** [comment]
Alright, thanks for the clarification @btmills.

**Comment by IanVS** [comment]
I just ran across this concept yesterday in: https://leanpub.com/exploring-es6/read#ch_modules.  It was surprising to me as well.

**Comment by nzakas** [comment]
Another good source: https://leanpub.com/understandinges6/read#leanpub-auto-modules

;)

**Comment by IanVS** [comment]
Oh man, how did I not know about this?  Definitely going to grab a copy.  :)
